### PR TITLE
Added http api

### PR DESCRIPTION
**Pull request by qusma**
Alright, so here's my initial take on the HTTP API. It's got more-or-less all the necessary "infrastructure", but not all the implementations. 

**Server**
CustomBootstrapper.cs sets up the http server.

I have implemented 6 endpoints so far (these are in QDMS.Server/NancyModules):

 * /exchanges
 * /instruments (replaced InstrumentServer)
 * /datasources
 * /tags
 * /sessiontemplates
 * /economicreleases

This also includes validation (in QDMS.Server/Validation). These validation classes can also be reused as they are for validation in the UI.

UI: I haven't done much in terms of starting to use the http api in the UI yet, in case we decide to make any major changes to it. I wrote an interface for tags as a proof of concept (see TagsWindow/TagsViewModel/TagViewModel). ValidatingViewModelBase provides what's needed to use the validation classes with the WPF UI.

Testing: I wrote exhaustive tests for ExchangeModule just for the hell of it, with more targeted coverage (on the parts likely to break) for the others. The infrastructure is in ModuleTestBase.

Two new settings added: API key and the http server port.

**Client**
All the url construction/headers/serialization/deserialization/error handling/etc is done in ApiClient.cs. 

This is then used in QDMSClient.Http.cs. Since everything is handled in ApiClient, the functions in QDMSClient are just one-liners where you specify 1) the path, 2) request method, and optionally 3) the type of object expected in response and 4) the object to send.

### CODE REVIEW

File: QDMSClient/ApiClient.cs [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+    /// Wraps around HttpClient with various useful functions for consuming a REST API
+    /// </summary>
+    public class ApiClient : IDisposable
+    {
+        private HttpClient _httpClient = new HttpClient();
+        private readonly string _baseAddr;
+
+        public ApiClient(string host, int httpPort, string apiKey)
+        {
+            _baseAddr = $"http://{host}:{httpPort}";
```
Comment (leo90skk) [inline QDMSClient/ApiClient.cs]: Would be nice to have the possibility to also create a client/server with https....

Comment (leo90skk) [inline QDMSClient/ApiClient.cs]: Would be nice to have the possibility to also create a client/server with https....

File: QDMSService/Config/RestService.cs [hunk 2]
```diff
@@ -7,14 +7,21 @@
 
 namespace QDMSService.Config
 {
-    public class InstrumentService : ConfigurationElement
+    public class RestService : ConfigurationElement
```
Comment (leo90skk) [inline QDMSService/Config/RestService.cs]: Shouldn't we call that class WebService?

File: QDMSTest/HttpServer/ExchangeModuleTest.cs [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+using Newtonsoft.Json;
+using NUnit.Framework;
+using QDMS;
+using QDMS.Server.NancyModules;
+using System;
+using System.Collections.Generic;
+using System.Data.Entity;
+using System.Linq;
+
+namespace QDMSTest.httpServer
```
Comment (leo90skk) [inline QDMSTest/HttpServer/ExchangeModuleTest.cs]: Can we make this uppercase at the start after Microsoft development guideline https://msdn.microsoft.com/en-us/library/x2dbyw72(v=vs.71).aspx ?

File: QDMSTest/HttpServer/ModuleTestBase.cs [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+using QDMSServer;
+using System;
+using System.Collections.Generic;
+using System.Data.Entity;
+using System.Data.Entity.Infrastructure;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace QDMSTest.httpServer
```
Comment (leo90skk) [inline QDMSTest/HttpServer/ModuleTestBase.cs]: here the same...

**Comment by qusma** [comment]
The reason for ApiResponse is that there's a gazillion things that can go wrong with these requests. Request something that doesn't exist, try to add something that breaks unique constraints, do something that does not pass validation, auth key is wrong, etc. It's also not always a problem you can describe with an HttpStatusCode, e.g. if the request times out the client will throw TaskCanceledException.

So you have to do error handling for every request. It's either something like ApiResponse, or throw exceptions, and I think the first approach is cleaner. It even lets you completely ignore error handling if you're so inclined.

```
var instrumentResponse = client.GetInstrument(5);
if (!instrumentResponse .WasSuccessful)
{
    //handle errors..
}
```

```
Instrument instrument;
try
{
    instrument = client.GetInstrument(5);
}
catch (Exception)
{
    //handle errors...
}
```

With C# 7 this could be made a bit nicer by returning a tuple (T, ApiResponse), so you wouldn't need to manually get the ApiResponse.Result object:
```
(instrument, response) = client.GetInstrument(5);
```

Added SSL option and wrote a little guide to setting up SSL: https://github.com/leo90skk/qdms/wiki/Setting-up-SSL

**Pull request merged**